File: examples/python/hypermachine_client.py
#!/usr/bin/env python3
"""
HyperMachine Python Client

A simple Python client for interacting with the HyperMachine MCP HTTP server.
Compatible with OpenAI function calling and direct API usage.

Usage:
    # Direct API usage
    client = HyperMachineClient("http://localhost:8080", api_key="your-key")
    
    # List tools (for OpenAI function calling)
    tools = client.list_tools()
    
    # Create a VM
    vm = client.create_vm("my-vm", cpu_cores=4, memory_gb=8)
    
    # Start the VM
    client.start_vm("my-vm")
    
    # Get VM status
    status = client.get_vm("my-vm")
    print(status)
"""

import json
import requests
from typing import Any, Optional
from dataclasses import dataclass


@dataclass
class VmInfo:
    """Virtual machine information."""
    name: str
    cpu_cores: int
    memory_gb: int
    gpu_enabled: bool
    network_enabled: bool
    state: str
    created_at: str
    updated_at: str


class HyperMachineError(Exception):
    """Exception raised by HyperMachine client."""
    def __init__(self, message: str, status_code: int = None):
        self.message = message
        self.status_code = status_code
        super().__init__(self.message)


class HyperMachineClient:
    """Client for the HyperMachine MCP HTTP server."""
    
    def __init__(self, base_url: str = "http://localhost:8080", api_key: str = None):
        """
        Initialize the HyperMachine client.
        
        Args:
            base_url: The base URL of the MCP server (default: http://localhost:8080)
            api_key: Optional API key for authentication
        """
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.session = requests.Session()
        
        if api_key:
            self.session.headers["Authorization"] = f"Bearer {api_key}"
        self.session.headers["Content-Type"] = "application/json"
    
    def _request(self, method: str, path: str, data: dict = None) -> dict:
        """Make an HTTP request to the server."""
        url = f"{self.base_url}{path}"
        
        try:
            if method == "GET":
                response = self.session.get(url)
            elif method == "POST":
                response = self.session.post(url, json=data)
            elif method == "DELETE":
                response = self.session.delete(url)
            else:
                raise ValueError(f"Unsupported method: {method}")
            
            if response.status_code == 401:
                raise HyperMachineError("Unauthorized - check your API key", 401)
            elif response.status_code == 429:
                raise HyperMachineError("Rate limit exceeded", 429)
            elif response.status_code >= 400:
                error_msg = response.text or f"HTTP {response.status_code}"
                raise HyperMachineError(error_msg, response.status_code)
            
            return response.json()
        except requests.RequestException as e:
            raise HyperMachineError(f"Request failed: {e}")
    
    # =========================================================================
    # Agentic AI Interface (for LLM integration)
    # =========================================================================
    
    def get_ontology(self) -> dict:
        """
        Get the complete HyperMachine ontology for AI agent discovery.
        
        Returns:
            Full ontology with concepts, types, operations, and relationships
        """
        return self._request("GET", "/agentic/ontology")
    
    def get_capabilities(self) -> dict:
        """
        Get a quick summary of available operations.
        
        Returns:
            Capabilities summary with operation list
        """
        return self._request("GET", "/agentic/capabilities")
    
    def get_schema(self, compact: bool = False) -> dict:
        """
        Get the JSON Schema for validation.
        
        Args:
            compact: If True, return minimal schema for bandwidth-constrained scenarios
            
        Returns:
            JSON Schema document
        """
        path = "/agentic/schema/compact" if compact else "/agentic/schema"
        return self._request("GET", path)
    
    def get_provider_config(self, provider: str) -> dict:
        """
        Get provider-specific configuration for an LLM.
        
        Args:
            provider: Provider name (openai, gpt-5, claude, claude-4.5, gemini, gemini-2.5, etc.)
            
        Returns:
            Configuration with tools, system_prompt, and hints
        """
        return self._request("GET", f"/agentic/providers/{provider}")
    
    def get_openai_tools(self) -> list[dict]:
        """
        Get tools in OpenAI function calling format.
        
        Supports: GPT-5, GPT-5-turbo, GPT-4o, o1, o3, etc.
        
        Returns:
            List of tools in OpenAI format
        """
        return self._request("GET", "/agentic/tools/openai")
    
    def get_anthropic_tools(self) -> list[dict]:
        """
        Get tools in Anthropic tool use format.
        
        Supports: Claude 4.5, Claude 4, Claude 3.5, Sonnet, Opus, Haiku
        
        Returns:
            List of tools in Anthropic format
        """
        return self._request("GET", "/agentic/tools/anthropic")
    
    def get_gemini_tools(self) -> dict:
        """
        Get tools in Google Gemini format.
        
        Supports: Gemini 2.5, Gemini 2.0, Gemini Flash
        
        Returns:
            Tools in Gemini function declaration format
        """
        return self._request("GET", "/agentic/tools/gemini")
    
    # =========================================================================
    # Tool Discovery (legacy MCP endpoint)
    # =========================================================================
    
    def list_tools(self) -> list[dict]:
        """
        Get available tools in basic format.
        
        Note: Prefer get_openai_tools() or get_anthropic_tools() for LLM integration.
        
        Returns:
            List of tool definitions with name, description, and parameters
        """
        return self._request("GET", "/mcp/tools")
    
    def list_tools_openai_format(self) -> list[dict]:
        """
        Get tools formatted for OpenAI's function calling API.
        
        Note: Prefer get_openai_tools() for optimized LLM integration.
        
        Returns:
            List of tools in OpenAI's expected format
        """
        tools = self.list_tools()
        return [
            {
                "type": "function",
                "function": {
                    "name": tool["name"],
                    "description": tool["description"],
                    "parameters": tool["parameters"]
                }
            }
            for tool in tools
        ]
    
    def call_tool(self, tool_name: str, arguments: dict) -> dict:
        """
        Execute a tool call (MCP unified endpoint).
        
        Args:
            tool_name: The tool to call (e.g., "vm.create", "vm.start")
            arguments: Tool-specific arguments
            
        Returns:
            Tool execution result
        """
        return self._request("POST", "/mcp/call", {
            "tool": tool_name,
            "arguments": arguments
        })
    
    # =========================================================================
    # VM Management (REST API)
    # =========================================================================
    
    def list_vms(self) -> list[VmInfo]:
        """List all virtual machines."""
        vms = self._request("GET", "/vms")
        return [VmInfo(**vm) for vm in vms]
    
    def create_vm(
        self,
        name: str,
        cpu_cores: int = 2,
        memory_gb: int = 4,
        gpu_enabled: bool = False,
        network_enabled: bool = False
    ) -> VmInfo:
        """
        Create a new virtual machine.
        
        Args:
            name: Unique VM name
            cpu_cores: Number of virtual CPUs (default: 2)
            memory_gb: Memory size in GB (default: 4)
            gpu_enabled: Enable GPU passthrough (default: False)
            network_enabled: Enable networking (default: False)
            
        Returns:
            Created VM information
        """
        data = {
            "name": name,
            "cpu_cores": cpu_cores,
            "memory_gb": memory_gb,
            "gpu_enabled": gpu_enabled,
            "network_enabled": network_enabled
        }
        result = self._request("POST", "/vms", data)
        return VmInfo(**result)
    
    def get_vm(self, name: str) -> VmInfo:
        """Get details of a specific VM."""
        result = self._request("GET", f"/vms/{name}")
        return VmInfo(**result)
    
    def delete_vm(self, name: str) -> dict:
        """Delete a virtual machine."""
        return self._request("DELETE", f"/vms/{name}")
    
    def start_vm(self, name: str) -> dict:
        """Start a virtual machine."""
        return self._request("POST", f"/vms/{name}/start")
    
    def stop_vm(self, name: str) -> dict:
        """Stop a virtual machine."""
        return self._request("POST", f"/vms/{name}/stop")
    
    def get_metrics(self, name: str) -> dict:
        """Get VM metrics (CPU, memory usage, uptime)."""
        return self._request("GET", f"/vms/{name}/metrics")
    
    def execute_script(self, name: str, script: str, timeout: int = 300) -> dict:
        """
        Execute a script inside a running VM.
        
        Args:
            name: VM name
            script: Script content to execute
            timeout: Execution timeout in seconds (default: 300)
            
        Returns:
            Script execution result
        """
        return self._request("POST", f"/vms/{name}/script", {
            "script": script,
            "timeout_seconds": timeout
        })
    
    def health_check(self) -> dict:
        """Check server health status."""
        return self._request("GET", "/health")


# =============================================================================
# OpenAI Integration Example
# =============================================================================

def openai_agent_example():
    """
    Example of using HyperMachine with OpenAI function calling.
    
    Requires: pip install openai
    """
    try:
        import openai
    except ImportError:
        print("Install openai package: pip install openai")
        return
    
    # Initialize clients
    hm = HyperMachineClient(api_key="your-api-key")
    
    # Get tools in OpenAI format
    tools = hm.list_tools_openai_format()
    
    # Create chat completion with tools
    response = openai.chat.completions.create(
        model="gpt-4",
        messages=[
            {"role": "user", "content": "Create a VM named 'web-server' with 4 CPU cores and 8GB RAM, then start it"}
        ],
        tools=tools,
    )
    
    # Process tool calls from the response
    for tool_call in response.choices[0].message.tool_calls or []:
        result = hm.call_tool(
            tool_call.function.name,
            json.loads(tool_call.function.arguments)
        )
        print(f"Tool: {tool_call.function.name}")
        print(f"Result: {result}")


# =============================================================================
# Direct Usage Example
# =============================================================================

def direct_usage_example():
    """Example of direct API usage."""
    # Initialize client (no auth for local development)
    client = HyperMachineClient("http://localhost:8080")
    
    # Check health
    health = client.health_check()
    print(f"Server status: {health['status']}")
    
    # Create a VM
    vm = client.create_vm(
        name="example-vm",
        cpu_cores=2,
        memory_gb=4,
        network_enabled=True
    )
    print(f"Created VM: {vm.name} ({vm.state})")
    
    # Start the VM
    client.start_vm("example-vm")
    print("VM started")
    
    # Get metrics
    metrics = client.get_metrics("example-vm")
    print(f"CPU usage: {metrics.get('cpu_percent', 0)}%")
    print(f"Memory usage: {metrics.get('memory_percent', 0)}%")
    
    # List all VMs
    vms = client.list_vms()
    print(f"\nAll VMs ({len(vms)}):")
    for v in vms:
        print(f"  - {v.name}: {v.state} ({v.cpu_cores} cores, {v.memory_gb} GB)")
    
    # Stop and delete
    client.stop_vm("example-vm")
    client.delete_vm("example-vm")
    print("\nVM stopped and deleted")


if __name__ == "__main__":
    print("HyperMachine Python Client")
    print("=" * 40)
    print("\nTo use this client:")
    print("  1. Start the MCP server: hm serve --rest-port 8080")
    print("  2. Run: python hypermachine_client.py")
    print("\nExample usage:")
    print("  client = HyperMachineClient('http://localhost:8080')")
    print("  client.create_vm('my-vm', cpu_cores=4, memory_gb=8)")
    print("  client.start_vm('my-vm')")

File: examples/typescript/hypermachine-client.ts
/**
 * HyperMachine TypeScript Client
 *
 * A TypeScript/JavaScript client for interacting with the HyperMachine MCP HTTP server.
 * Compatible with OpenAI function calling and direct API usage.
 *
 * @example
 * ```typescript
 * const client = new HyperMachineClient("http://localhost:8080", "your-api-key");
 *
 * // Create a VM
 * const vm = await client.createVm("my-vm", { cpuCores: 4, memoryGb: 8 });
 *
 * // Start it
 * await client.startVm("my-vm");
 *
 * // Get tools for OpenAI
 * const tools = await client.listToolsOpenAIFormat();
 * ```
 */

export interface VmInfo {
    name: string;
    cpu_cores: number;
    memory_gb: number;
    gpu_enabled: boolean;
    network_enabled: boolean;
    state: string;
    created_at: string;
    updated_at: string;
}

export interface ToolDefinition {
    name: string;
    description: string;
    parameters: Record<string, unknown>;
}

export interface OpenAITool {
    type: "function";
    function: {
        name: string;
        description: string;
        parameters: Record<string, unknown>;
    };
}

export interface ToolCallResult {
    success: boolean;
    result?: unknown;
    error?: string;
}

export interface VmCreateOptions {
    cpuCores?: number;
    memoryGb?: number;
    gpuEnabled?: boolean;
    networkEnabled?: boolean;
}

export interface VmMetrics {
    cpu_percent: number;
    memory_percent: number;
    uptime_seconds: number;
}

export class HyperMachineError extends Error {
    constructor(
        message: string,
        public statusCode?: number
    ) {
        super(message);
        this.name = "HyperMachineError";
    }
}

export class HyperMachineClient {
    private baseUrl: string;
    private apiKey?: string;

    /**
     * Create a new HyperMachine client.
     *
     * @param baseUrl - The base URL of the MCP server (default: http://localhost:8080)
     * @param apiKey - Optional API key for authentication
     */
    constructor(baseUrl: string = "http://localhost:8080", apiKey?: string) {
        this.baseUrl = baseUrl.replace(/\/$/, "");
        this.apiKey = apiKey;
    }

    private async request<T>(
        method: string,
        path: string,
        body?: unknown
    ): Promise<T> {
        const headers: Record<string, string> = {
            "Content-Type": "application/json",
        };

        if (this.apiKey) {
            headers["Authorization"] = `Bearer ${this.apiKey}`;
        }

        const response = await fetch(`${this.baseUrl}${path}`, {
            method,
            headers,
            body: body ? JSON.stringify(body) : undefined,
        });

        if (response.status === 401) {
            throw new HyperMachineError("Unauthorized - check your API key", 401);
        }
        if (response.status === 429) {
            throw new HyperMachineError("Rate limit exceeded", 429);
        }
        if (!response.ok) {
            const text = await response.text();
            throw new HyperMachineError(
                text || `HTTP ${response.status}`,
                response.status
            );
        }

        return response.json();
    }

    // ===========================================================================
    // Agentic AI Interface (for LLM integration)
    // ===========================================================================

    /**
     * Get the complete HyperMachine ontology for AI agent discovery.
     */
    async getOntology(): Promise<Record<string, unknown>> {
        return this.request<Record<string, unknown>>("GET", "/agentic/ontology");
    }

    /**
     * Get a quick summary of available operations.
     */
    async getCapabilities(): Promise<Record<string, unknown>> {
        return this.request<Record<string, unknown>>("GET", "/agentic/capabilities");
    }

    /**
     * Get the JSON Schema for validation.
     * @param compact - If true, return minimal schema for bandwidth-constrained scenarios
     */
    async getSchema(compact: boolean = false): Promise<Record<string, unknown>> {
        const path = compact ? "/agentic/schema/compact" : "/agentic/schema";
        return this.request<Record<string, unknown>>("GET", path);
    }

    /**
     * Get provider-specific configuration for an LLM.
     * @param provider - Provider name (openai, gpt-5, claude, claude-4.5, gemini, gemini-2.5, etc.)
     */
    async getProviderConfig(provider: string): Promise<{
        provider: string;
        tools: unknown;
        system_prompt: string;
        hints: {
            temperature: number;
            parallel_tool_calls: boolean;
            max_tokens?: number;
        };
    }> {
        return this.request("GET", `/agentic/providers/${encodeURIComponent(provider)}`);
    }

    /**
     * Get tools in OpenAI function calling format.
     * Supports: GPT-5, GPT-5-turbo, GPT-4o, o1, o3, etc.
     */
    async getOpenAITools(): Promise<OpenAITool[]> {
        return this.request<OpenAITool[]>("GET", "/agentic/tools/openai");
    }

    /**
     * Get tools in Anthropic tool use format.
     * Supports: Claude 4.5, Claude 4, Claude 3.5, Sonnet, Opus, Haiku
     */
    async getAnthropicTools(): Promise<Array<{
        name: string;
        description: string;
        input_schema: Record<string, unknown>;
    }>> {
        return this.request("GET", "/agentic/tools/anthropic");
    }

    /**
     * Get tools in Google Gemini format.
     * Supports: Gemini 2.5, Gemini 2.0, Gemini Flash
     */
    async getGeminiTools(): Promise<{
        function_declarations: Array<{
            name: string;
            description: string;
            parameters: Record<string, unknown>;
        }>;
    }> {
        return this.request("GET", "/agentic/tools/gemini");
    }

    // ===========================================================================
    // Tool Discovery (legacy MCP endpoint)
    // ===========================================================================

    /**
     * Get available tools in MCP format.
     * Note: Prefer getOpenAITools() or getAnthropicTools() for LLM integration.
     */
    async listTools(): Promise<ToolDefinition[]> {
        return this.request<ToolDefinition[]>("GET", "/mcp/tools");
    }

    /**
     * Get tools formatted for OpenAI's function calling API.
     * Note: Prefer getOpenAITools() for optimized LLM integration.
     */
    async listToolsOpenAIFormat(): Promise<OpenAITool[]> {
        const tools = await this.listTools();
        return tools.map((tool) => ({
            type: "function" as const,
            function: {
                name: tool.name,
                description: tool.description,
                parameters: tool.parameters,
            },
        }));
    }

    /**
     * Execute a tool call via the MCP unified endpoint.
     */
    async callTool(toolName: string, args: unknown): Promise<ToolCallResult> {
        return this.request<ToolCallResult>("POST", "/mcp/call", {
            tool: toolName,
            arguments: args,
        });
    }

    // ===========================================================================
    // VM Management (REST API)
    // ===========================================================================

    /**
     * List all virtual machines.
     */
    async listVms(): Promise<VmInfo[]> {
        return this.request<VmInfo[]>("GET", "/vms");
    }

    /**
     * Create a new virtual machine.
     */
    async createVm(name: string, options: VmCreateOptions = {}): Promise<VmInfo> {
        return this.request<VmInfo>("POST", "/vms", {
            name,
            cpu_cores: options.cpuCores ?? 2,
            memory_gb: options.memoryGb ?? 4,
            gpu_enabled: options.gpuEnabled ?? false,
            network_enabled: options.networkEnabled ?? false,
        });
    }

    /**
     * Get details of a specific VM.
     */
    async getVm(name: string): Promise<VmInfo> {
        return this.request<VmInfo>("GET", `/vms/${encodeURIComponent(name)}`);
    }

    /**
     * Delete a virtual machine.
     */
    async deleteVm(name: string): Promise<{ status: string; name: string }> {
        return this.request("DELETE", `/vms/${encodeURIComponent(name)}`);
    }

    /**
     * Start a virtual machine.
     */
    async startVm(name: string): Promise<{ status: string; name: string }> {
        return this.request("POST", `/vms/${encodeURIComponent(name)}/start`);
    }

    /**
     * Stop a virtual machine.
     */
    async stopVm(name: string): Promise<{ status: string; name: string }> {
        return this.request("POST", `/vms/${encodeURIComponent(name)}/stop`);
    }

    /**
     * Get VM metrics.
     */
    async getMetrics(name: string): Promise<VmMetrics> {
        return this.request<VmMetrics>(
            "GET",
            `/vms/${encodeURIComponent(name)}/metrics`
        );
    }

    /**
     * Execute a script inside a running VM.
     */
    async executeScript(
        name: string,
        script: string,
        timeoutSeconds: number = 300
    ): Promise<unknown> {
        return this.request(
            "POST",
            `/vms/${encodeURIComponent(name)}/script`,
            {
                script,
                timeout_seconds: timeoutSeconds,
            }
        );
    }

    /**
     * Check server health.
     */
    async healthCheck(): Promise<{ status: string; service: string; version: string }> {
        return this.request("GET", "/health");
    }
}

// =============================================================================
// OpenAI Integration Example
// =============================================================================

/**
 * Example: Using HyperMachine with OpenAI function calling.
 *
 * ```typescript
 * import OpenAI from 'openai';
 * import { HyperMachineClient } from './hypermachine-client';
 *
 * const hm = new HyperMachineClient("http://localhost:8080", "your-api-key");
 * const openai = new OpenAI();
 *
 * // Get tools
 * const tools = await hm.listToolsOpenAIFormat();
 *
 * // Chat with tools
 * const response = await openai.chat.completions.create({
 *   model: "gpt-4",
 *   messages: [{ role: "user", content: "Create a VM with 4 cores" }],
 *   tools,
 * });
 *
 * // Execute tool calls
 * for (const toolCall of response.choices[0].message.tool_calls || []) {
 *   const result = await hm.callTool(
 *     toolCall.function.name,
 *     JSON.parse(toolCall.function.arguments)
 *   );
 *   console.log(result);
 * }
 * ```
 */

// =============================================================================
// Direct Usage Example
// =============================================================================

async function example() {
    const client = new HyperMachineClient("http://localhost:8080");

    // Check health
    const health = await client.healthCheck();
    console.log(`Server: ${health.status}`);

    // Create a VM
    const vm = await client.createVm("example-vm", {
        cpuCores: 2,
        memoryGb: 4,
        networkEnabled: true,
    });
    console.log(`Created: ${vm.name} (${vm.state})`);

    // Start it
    await client.startVm("example-vm");

    // List all VMs
    const vms = await client.listVms();
    console.log(`Total VMs: ${vms.length}`);

    // Cleanup
    await client.stopVm("example-vm");
    await client.deleteVm("example-vm");
}

// Run if executed directly
if (typeof require !== "undefined" && require.main === module) {
    example().catch(console.error);
}

export default HyperMachineClient;
